### PR TITLE
Fix job counter + refactor TCP based queries

### PR DESCRIPTION
**Pull request by EmielBruijntjes**
There was an issue with the counters in Core.cpp because not all calls back to userspace were triggered by a Core-method-call (for example, tcp connections not). In this branche we refactored the TCP implementation to reuse the same TCP connection for all queries, and at the same time fix the counters inside Core.

### CODE REVIEW

File: src/remotelookup.cpp [hunk 1]
```diff
@@ -227,69 +235,74 @@ void RemoteLookup::report(const Response &response)
...
-    _connection.reset(new Connection(_core->loop(), ip, _query, response, this));
-    
     // we can unsubscribe from all inbound udp sockets because we're no longer interested in those responses
     unsubscribe();
     
+    // try to connect to a TCP socket
+    if (!_core->connect(ip, this)) return report(response);
+    
+    // because we switched to tcp mode, we remember the truncated response in case tcp fails too,
+    // we then at least have _something_ to report
```
Comment (rwols) [inline src/remotelookup.cpp]: I think it'd be better to resolve with a failure than resolving with data that's incomplete, but suggested as a total success.

File: include/dnscpp/tcp.h [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     *  @return size_t
+     */
+    virtual size_t deliver(size_t maxcalls) override;
+};
+    
+/**
+ *  End of namespace
+ */
+}
+
```
Comment (rwols) [inline include/dnscpp/tcp.h]: Redundant empty line.

File: include/dnscpp/tcp.h [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+     *  The filedescriptor
+     *  @var int
+     */
+    int _fd;
+    
+    /**
+     *  The buffer that is being filled right now (first two bytes contain the size)
+     *  @var unsigned char *
+     */
+    unsigned char *_buffer = nullptr;
```
Comment (rwols) [inline include/dnscpp/tcp.h]: We're not in the 90's anymore. Use something like `std::vector<unsigned char>` to manage a sequence of bytes.

File: src/locallookup.h [hunk 4]
```diff
@@ -74,13 +71,34 @@ class LocalLookup : public Lookup
...
+        // handler is reset on completion
+        return _handler == nullptr;
+    }
+    
+    /**
+     *  Is this lookup exhausted: meaning that it has sent its max number of requests, but still
+     *  has not received an appropriate answer, and is now waiting for its final timer to finish
+     *  @return bool
      */
-    virtual size_t credits() const override
```
Comment (rwols) [inline src/locallookup.h:80]: It's good to see this gone as the method doesn't tell me what it's doing at all.

File: src/udp.cpp [hunk 5]
```diff
@@ -178,82 +184,12 @@ void Udp::notify()
...
-            // @todo make this more efficient (without all the string-copying)
-            _responses.emplace_back(reinterpret_cast<const sockaddr*>(&from), std::basic_string<unsigned char>(buffer, bytes));
-        }
-        catch (...)
-        {
-            // ip address could not be parsed
-        }
+        
+        // add to the buffer
+        add((const sockaddr *)&from, buffer, bytes);
```
Comment (rwols) [inline src/udp.cpp:189]: I don't think this needs a separate method. It's pretty straightforward what happens.

File: include/dnscpp/sockets.h [hunk 6]
```diff
@@ -55,12 +58,12 @@ class Udps : private Watchable, Udp::Handler
...
         /**
-         *  Method that is called when the udp socket has a buffer available
+         *  Method that is called when one of the sockets has a buffer of received messages
          *  @param  udp     the reporting socket
          */
-        virtual void onBuffered(Udps *udp) = 0;
+        virtual void onBuffered(Sockets *udp) = 0;
     };
-
+
```
Comment (rwols) [inline include/dnscpp/sockets.h:66]: Redundant spaces.

File: src/locallookup.h [hunk 7]
```diff
@@ -36,30 +36,27 @@ class LocalLookup : public Lookup
...
+     *  @return bool    was there a call to userspace?
      */
     virtual bool execute(double now) override
     {
         // do nothing if ready
-        if (_ready) return false;
-        
-        // pass to the hosts
-        _hosts.notify(Request(this), _handler, this);
+        if (_handler == nullptr) return false;
```
Comment (rwols) [inline src/locallookup.h]:
```suggestion
        if (finished()) return false;
```

In fact, I think this should be asserted: `assert(!finished())` as it seems like a programming error to not check this before calling execute.

**Review by rwols — CHANGES_REQUESTED** [review]

File: src/udp.cpp (hunk 5 above)
Comment (EmielBruijntjes) [reply src/udp.cpp:189]: The method is trivial indeed, but we need it to get access to a private data structure in the base class

File: src/locallookup.h (hunk 7 above)
Comment (EmielBruijntjes) [reply src/locallookup.h]: I find this difficult. It is true that from the perspective of the entire library this method should indeed not be called when the lookup is already finished. But from the perspective of just this class (or just this method) we prepare for the worst and also handle cases that there is no crash even when it _is_ called...

**Pull request merged**